Implement mv method in GDriveFileSystem

simone-viozzi opened the pull request:
Hi, i'm addind a method to rename a `GoogleDriveFile`

the google api to rename a file is: https://developers.google.com/drive/api/v2/reference/files/patch

I saw you have a wrapper for `files().patch`: `_FilesPatch`
do you prefer if I use the existing wrapper or the api directly?

shcheklein commented:
@simone-viozzi this is from the docs:

`Manipulate file metadata and contents from [GoogleDriveFile](https://docs.iterative.ai/PyDrive2/pydrive2/#pydrive2.files.GoogleDriveFile) object and call [Upload()](https://docs.iterative.ai/PyDrive2/pydrive2/#pydrive2.files.GoogleDriveFile.Upload)`

Does it make sense to use the Upload for this?

E.g. like this https://docs.iterative.ai/PyDrive2/filemanagement/?highlight=metadata#update-file-metadata

shcheklein reviewed (changes requested):
Need to review the workflow around `Upload()`

simone-viozzi commented:
Right now is possible to rename a file using existing methods:

```python
file.FetchMetadata(fields="title")
file["title"] = "new_file1"
file.Upload()
```

and looking at the code for `upload`:
https://github.com/iterative/PyDrive2/blob/64a50a1580b2d276f8c9563f5ba620f7e7706119/pydrive2/files.py#L499-L512

it uses `_FilesPatch` internally.

i discovered this way to rename a file with https://github.com/iterative/PyDrive2/issues/91

but it would be much more convenient and user friendly to just have a method that does that

shcheklein commented:
@simone-viozzi could you check if fsspec interface had `move` / `rename` please? May be we should implement it there (keep this one as-is - low level). 

one more question:

```
file.FetchMetadata(fields="title")
file["title"] = "new_file1"
file.Upload()
```

is it possible to do this w/o a call to Fetch? (if I know the file id). I wonder also if it's possible to do this if I know only a file name w/o any extra calls. That would be a good benefit of this I think.

simone-viozzi commented:
> is it possible to do this w/o a call to Fetch? (if I know the file id). I wonder also if it's possible to do this if I know only a file name w/o any extra calls. That would be a good benefit of this I think.

No, without fetch it doesn't do nothing. 

Consider the minimal example:

```python
file = drive.CreateFile({'id': "13nYjSBX7FmSUFnBcTuT6hKsFswXhqcpzpioei5r2eRQ"})
file["title"] = "new_file1"
file.Upload()
```

The stack-trace of `Upload` in this case is:
- `Upload`
- `_FilesPatch`
- `GetChanges` -> this return a blank dict

https://github.com/iterative/PyDrive2/blob/64a50a1580b2d276f8c9563f5ba620f7e7706119/pydrive2/apiattr.py#L92-L103

i think because the title was `None` and it doesn't count as a change, but i might be wrong on this one. 

With the rename method i implemented: 

https://github.com/simone-viozzi/PyDrive2/blob/c55271427ace18d59019f03392e73cf99d46f9eb/pydrive2/files.py#L574-L587

you don't need to fetch it before because i specified it in the request: `param["fields"] = "title"`


> @simone-viozzi could you check if fsspec interface had move / rename please? May be we should implement it there (keep this one as-is - low level).

there should be a move method (https://github.com/iterative/PyDrive2/pull/119) but i can't find it on the current version. Also there is a reference to it in one of the tests:

https://github.com/iterative/PyDrive2/blob/64a50a1580b2d276f8c9563f5ba620f7e7706119/pydrive2/test/test_fs.py#L90-L102

i suppose it's an old test and the method was dropped somewhere.

Anyway, there is no `rename` / `move` on the current version of fsspec

shcheklein commented:
> i discovered this way to rename a file with https://github.com/iterative/PyDrive2/issues/91

yep, thanks. I remember it now. I was trying to fix it, but it was not that trivial. Ideally, it should be fixed though.

> there should be a move method (https://github.com/iterative/PyDrive2/pull/119) but i can't find it on the current version. Also there is a reference to it in one of the tests:

https://filesystem-spec.readthedocs.io/en/latest/api.html#fsspec.spec.AbstractFileSystem.rename

That's the default implementation of `move`:

https://github.com/fsspec/filesystem_spec/blob/master/fsspec/spec.py#L917-L920

That's the `rename`:

https://github.com/fsspec/filesystem_spec/blob/master/fsspec/spec.py#L1251

It's not terrible (if we optimize `copy`). But both can improved to use a single API call I think.

---------

So, ideally, we can take a look and see if we can fix the #91 + optimize fsspec to provide a better interface for this. Then we can see if we need a separate wrapper on top of `_Patch`

Let me know wdyt.

simone-viozzi commented:
oh, i get it now, thank you! 

i think we can definitively optimize them, maybe for the rename we could use `self.auth.service.files().patch(**param).execute()` directly, so we don't need the fetch before the call.
But this way we would need a function on the `GoogleDriveFile` level and another on the `GDriveFileSystem` level.

otherwise we can do just a function on the `GDriveFileSystem` level like this:

```python
def rename(path, new_title)
    id = _get_item_id(path)
    file = drive.CreateFile({'id': id})
    file.FetchMetadata(fields="title")
    file["title"] = new_title
    file.Upload()
```

But the current implementation of rename is: `def rename(self, path1, path2, **kwargs):` and i don't think we can change the signature.

So maybe the best is to change the implementation of move in a way that:

- if only the last segment of the path is changed it will rename the file
- if other segments are changed too it will move the file

This way we can leave the default implementation of rename.

The move method will become something like:

```python
def move(self, path1, path2):
    path1_left = path1.split('/')[:-1]
    path1_right = path1.split('/')[-1]

    path2_left = path1.split('/')[:-1]
    path2_right = path1.split('/')[-1]

   if path1_left == path2_left:
        rename file from path1_right to path2_right
   else:
        move file from path1 to path2
```

what do you think?

This way even if someone use move to rename a file it would call the right API

shcheklein commented:
Re moving files, check this https://developers.google.com/drive/api/guides/folder#move_files_between_folders

I feel it can be generalized to a single Patch call that updates parents + title if needed, wdyt? The only thing we'll need is to get ids for those parents, for the file, etc. It should be done carefully. The most annoying thing on Google Drive is that one path can resolve into multiple IDs.

simone-viozzi commented:
> The most annoying thing on Google Drive is that one path can resolve into multiple IDs.

check out: https://rclone.org/commands/rclone_dedupe/

i think for the move / rename method we can just throw an error if a path resolve in multiple IDs.

> Re moving files, check this https://developers.google.com/drive/api/guides/folder#move_files_between_folders

Good! 
so we need `patch` to rename the file and `update` to move it.
we can use `GoogleDriveFile.Upload` that does both, i will try it and get back

simone-viozzi commented:
mh, funny. 
The rename part does work:

```python
fs.mv('root/tmp/a.pdf', 'root/tmp/b.pdf')
```

but the move part does not.

Even if i change the parent array, the `Upload` method will call `_FilesPatch` instead of `_FilesUpdate`.

https://github.com/iterative/PyDrive2/blob/64a50a1580b2d276f8c9563f5ba620f7e7706119/pydrive2/files.py#L499-L512

I also tried on manually calling the privates methods:

```python
if file1_name != file2_name:
    file1._FilesPatch()

if file1_parent != file2_parent:
    file1._FilesUpdate()
```

But `param["body"] = self.GetChanges()` in `_FilesUpdate` does not recognize that the parent array has changed and nothing get updated.

@shcheklein do you have any idea why?

if i did the fetch before it should get that the array has changed

simone-viozzi commented:
Ok, now it does work.

both rename / move and rename and move at the same time.

i changed:

```python
file1["parents"].append({"id": file2_parent_id})
```

to:

```python
file1["parents"] = [{"id": file2_parent_id}]
```

This way when `_FilesUpdate` get the changes in the dict, the list is different ( different object ) and it gets updated.

i don't know if it's ok to just reset the parents array, on a normal drive it's not allowed to have a folder with multiple parents but in some other drives maybe is possible.

simone-viozzi commented:
I also discovered that we don't need to use `_FilesPatch`, but `_FilesUpdate` can also rename the file.

Please take a look at the code, for a single file it should be ready for a review.

shcheklein commented on pydrive2/fs/spec.py:
```diff
@@ -542,6 +542,34 @@ def cp_file(self, lpath, rpath, **kwargs):
             buffer = io.BytesIO(stream.read())
             self.upload_fobj(buffer, rpath)
 
+    def mv(self, path1, path2, recursive=False, maxdepth=None, **kwargs):
```
we probably need `_gdrive_retry` decorator here

shcheklein commented on pydrive2/fs/spec.py:
```diff
@@ -542,6 +542,34 @@ def cp_file(self, lpath, rpath, **kwargs):
+            file2_parent_id = self._get_item_id(file2_parent)
+            file1["parents"] = [{"id": file2_parent_id}]
+
+        file1._FilesUpdate()
```
after this operation is done we need to update or invalidate the ids cache `self._ids_cache` (see around how it's managed, probably we might need another helper that is wrapped with `@wrap_with(threading.RLock())`)

simone-viozzi replied:
sorry but i couldn't find any instance where the cache gets invalidated. 
even in `rm_file` the cache is not touched:

https://github.com/iterative/PyDrive2/blob/64a50a1580b2d276f8c9563f5ba620f7e7706119/pydrive2/fs/spec.py#L590-L605

how do i remove an element from the cache?

shcheklein replied:
resolving this since the whole cache mechanism is quite broken in GDrive's fsspec. Let's add as a TODO here

shcheklein commented on pydrive2/fs/spec.py:
```diff
@@ -542,6 +542,34 @@ def cp_file(self, lpath, rpath, **kwargs):
+        file1_name = posixpath.basename(path1)
+        file2_name = posixpath.basename(path2)
+
+        file1_parent = self._parent(path1)
```
Q / check: can `path2` be an absolute one, or is it still within the `self.path` directory. 

Do we want to support a way to move files outside, support absolute paths?

shcheklein commented on pydrive2/fs/spec.py:
```diff
@@ -542,6 +542,34 @@ def cp_file(self, lpath, rpath, **kwargs):
+            file1["title"] = file2_name
+
+        if file1_parent != file2_parent:
+            file2_parent_id = self._get_item_id(file2_parent)
```
Q / check: can path2 be an absolute one, or is it still within the self.path directory.

Do we want to support a way to move files outside, support absolute paths?

shcheklein replied:
check what happens if `file2_parent` is empty

simone-viozzi replied:
> Q / check: can path2 be an absolute one, or is it still within the self.path directory.
> Do we want to support a way to move files outside, support absolute paths?

```python
fs = GDriveFileSystem("root/tmp", auth)
fs.mv('root/tmp/a.pdf', 'root/test/b.pdf')
```

This worked. so we can move files even outside `self.path`.



> check what happens if file2_parent is empty

`file2_parent` cannot be empty otherwise `file2_parent_id = self._get_item_id(file2_parent)` will fail.
but i tried to force:
```python
file2_parent_id = None # or ""
file1["parents"] = [{"id": file2_parent_id}]
```
and i got an error:

```python
Traceback (most recent call last):
  File "/home/simone/workspace/drive-transfere-ownership/PyDrive2/pydrive2/files.py", line 780, in _FilesUpdate
    self.auth.service.files()
  File "/home/simone/workspace/drive-transfere-ownership/.direnv/python-3.8.12/lib/python3.8/site-packages/googleapiclient/_helpers.py", line 130, in positional_wrapper
    return wrapped(*args, **kwargs)
  File "/home/simone/workspace/drive-transfere-ownership/.direnv/python-3.8.12/lib/python3.8/site-packages/googleapiclient/http.py", line 938, in execute
    raise HttpError(resp, content, uri=self.uri)
googleapiclient.errors.HttpError: <HttpError 404 when requesting https://www.googleapis.com/drive/v2/files/1KSvXEOjUdDwTR2EfKm6Zl2xKX1OEmJgY?supportsAllDrives=true&alt=json returned "File not found:". Details: "[{'domain': 'global', 'reason': 'notFound', 'message': 'File not found: ', 'locationType': 'other', 'location': 'file'}]">

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/simone/workspace/drive-transfere-ownership/tests/gdrivefs.py", line 86, in <module>
    fs.mv('root/tmp/b.pdf', 'root/tmp/aa/a.pdf')
  File "/home/simone/workspace/drive-transfere-ownership/.direnv/python-3.8.12/lib/python3.8/site-packages/funcy/decorators.py", line 45, in wrapper
    return deco(call, *dargs, **dkwargs)
  File "/home/simone/workspace/drive-transfere-ownership/.direnv/python-3.8.12/lib/python3.8/site-packages/funcy/flow.py", line 127, in retry
    return call()
  File "/home/simone/workspace/drive-transfere-ownership/.direnv/python-3.8.12/lib/python3.8/site-packages/funcy/decorators.py", line 66, in __call__
    return self._func(*self._args, **self._kwargs)
  File "/home/simone/workspace/drive-transfere-ownership/PyDrive2/pydrive2/fs/spec.py", line 572, in mv
    file1._FilesUpdate()
  File "/home/simone/workspace/drive-transfere-ownership/PyDrive2/pydrive2/auth.py", line 84, in _decorated
    return decoratee(self, *args, **kwargs)
  File "/home/simone/workspace/drive-transfere-ownership/PyDrive2/pydrive2/files.py", line 56, in _decorated
    return decoratee(self, *args, **kwargs)
  File "/home/simone/workspace/drive-transfere-ownership/PyDrive2/pydrive2/files.py", line 785, in _FilesUpdate
    raise ApiRequestError(error)
pydrive2.files.ApiRequestError: <HttpError 404 when requesting https://www.googleapis.com/drive/v2/files/1KSvXEOjUdDwTR2EfKm6Zl2xKX1OEmJgY?supportsAllDrives=true&alt=json returned "File not found:". Details: "[{'domain': 'global', 'reason': 'notFound', 'message': 'File not found: ', 'locationType': 'other', 'location': 'file'}]">
```

so, `file2_parent_id` needs to be a valid id.

shcheklein replied:
Q: so, what is the usual interface for the other existing calls (e.g. remove, or ls, etc) - does it expect a relative path (w/o `root`) or an absolute one with `root`?

simone-viozzi replied:
I'm not sure, but i think everything want an absolute path.

fot example `ls` return `None` or get buggy with relative paths.

es:

```python
fs = GDriveFileSystem("root/tmp", auth)

# self.path
print(fs.ls("root/tmp"))
# > ['root/tmp/folder1/', [other files]]
print(fs.ls("root/tmp/folder1"))
# > ['root/tmp/folder1/folder2/', 'root/tmp/folder1/file1']
print(fs.ls("root/tmp/folder1/folder2"))
# > ['root/tmp/folder1/folder2/file2']

# self.path = root/tmp
# .
# |- folder1
# |  |- file1
# |  |- folder2
# |  |  |- file2

print(fs.ls("."))
# > ['./tmp/', './[files in my drive root]/']
print(fs.ls("./folder1"))
# > None
print(fs.ls("./folder1/folder2"))
# > None
print(fs.ls("folder1"))
# > ['folder1/tmp/', 'folder1/[files in my drive root]/']
print(fs.ls("folder1/folder2"))
# > None
```

shcheklein commented on pydrive2/fs/spec.py:
```diff
@@ -542,6 +542,34 @@ def cp_file(self, lpath, rpath, **kwargs):
+        file2_name = posixpath.basename(path2)
+
+        file1_parent = self._parent(path1)
+        file2_parent = self._parent(path2)
```
check: can any of those be None?

simone-viozzi replied:
if `file1_parent` is None, `file1_parent != file2_parent` is true and the function would try to move the file.

if `file2_parent` is None, `file2_parent_id = self._get_item_id(file2_parent)` will fail:

```python
Traceback (most recent call last):
  File "/home/simone/workspace/drive-transfere-ownership/tests/gdrivefs.py", line 87, in <module>
    fs.mv('root/tmp/test/b.pdf', 'root/tmp/a.pdf')
  File "/home/simone/workspace/drive-transfere-ownership/.direnv/python-3.8.12/lib/python3.8/site-packages/funcy/decorators.py", line 45, in wrapper
    return deco(call, *dargs, **dkwargs)
  File "/home/simone/workspace/drive-transfere-ownership/.direnv/python-3.8.12/lib/python3.8/site-packages/funcy/flow.py", line 127, in retry
    return call()
  File "/home/simone/workspace/drive-transfere-ownership/.direnv/python-3.8.12/lib/python3.8/site-packages/funcy/decorators.py", line 66, in __call__
    return self._func(*self._args, **self._kwargs)
  File "/home/simone/workspace/drive-transfere-ownership/PyDrive2/pydrive2/fs/spec.py", line 569, in mv
    file2_parent_id = self._get_item_id(file2_parent)
  File "/home/simone/workspace/drive-transfere-ownership/PyDrive2/pydrive2/fs/spec.py", line 367, in _get_item_id
    bucket, base = self.split_path(path)
  File "/home/simone/workspace/drive-transfere-ownership/PyDrive2/pydrive2/fs/spec.py", line 239, in split_path
    parts = path.replace("//", "/").rstrip("/").split("/", 1)
AttributeError: 'NoneType' object has no attribute 'replace'
```

if `file2_parent` is `""`, `file2_parent_id = self._get_item_id(file2_parent)` will fail:

```python
Traceback (most recent call last):
  File "/home/simone/workspace/drive-transfere-ownership/tests/gdrivefs.py", line 87, in <module>
    fs.mv('root/tmp/test/b.pdf', 'root/tmp/a.pdf')
  File "/home/simone/workspace/drive-transfere-ownership/.direnv/python-3.8.12/lib/python3.8/site-packages/funcy/decorators.py", line 45, in wrapper
    return deco(call, *dargs, **dkwargs)
  File "/home/simone/workspace/drive-transfere-ownership/.direnv/python-3.8.12/lib/python3.8/site-packages/funcy/flow.py", line 127, in retry
    return call()
  File "/home/simone/workspace/drive-transfere-ownership/.direnv/python-3.8.12/lib/python3.8/site-packages/funcy/decorators.py", line 66, in __call__
    return self._func(*self._args, **self._kwargs)
  File "/home/simone/workspace/drive-transfere-ownership/PyDrive2/pydrive2/fs/spec.py", line 569, in mv
    file2_parent_id = self._get_item_id(file2_parent)
  File "/home/simone/workspace/drive-transfere-ownership/PyDrive2/pydrive2/fs/spec.py", line 368, in _get_item_id
    assert bucket == self.root
AssertionError
```

shcheklein replied:
Q: do we need to handle this? I mean can the be actually None or empty?

simone-viozzi replied:
I don't think so, if it doesn't find the parent it will just throw a `FileNotFoundError: [Errno 2] No such file or directory:`

shcheklein commented on pydrive2/fs/spec.py:
```diff
@@ -542,6 +542,34 @@ def cp_file(self, lpath, rpath, **kwargs):
+
+        file1 = self.client.CreateFile({"id": file1_id})
+
+        file1.FetchMetadata(fields="title,parents")
```
need to fix it eventually ...  I don't think we have to do this call

simone-viozzi replied:
To fix this we need a way to tell `_FilesUpdate` which fields to include in the body. 

Now this in done trough `param["body"] = self.GetChanges()`:

line 770:
https://github.com/iterative/PyDrive2/blob/64a50a1580b2d276f8c9563f5ba620f7e7706119/pydrive2/files.py#L761-L789

and `GetChanges` is:

https://github.com/iterative/PyDrive2/blob/64a50a1580b2d276f8c9563f5ba620f7e7706119/pydrive2/apiattr.py#L92-L103

And that returns a blank dict if the metadata was not fetched before.

I still don't understand this logic well but i think that when you do `FetchMetadata`, at the end calls:

https://github.com/iterative/PyDrive2/blob/64a50a1580b2d276f8c9563f5ba620f7e7706119/pydrive2/apiattr.py#L86-L90

that is the same dict used by `GetChanges`.

If there is a bug, it's there somewhere. i will try to play around and see what happen.

simone-viozzi replied:
@shcheklein what do you think? i can't dig deeper than that, i don't really know the internal logic

shcheklein replied:
@simone-viozzi hey, sorry, I haven't had time to get to this :( we got into a few areas that would require some brainstorming, so I need a bit more time. I'll try to get there soon.

shcheklein replied:
https://github.com/iterative/PyDrive2/pull/223 should fix this, please rebase and let's try to simplify this

shcheklein commented:
@simone-viozzi thanks! it's a good iteration, I put some comments to improve it.

> i don't know if it's ok to just reset the parents array, on a normal drive it's not allowed to have a folder with multiple parents but in some other drives maybe is possible.

can you give more context please?

simone-viozzi commented:
> > i don't know if it's ok to just reset the parents array, on a normal drive it's not allowed to have a folder with multiple parents but in some other drives maybe is possible.

> can you give more context please?

if we print the metadata of a file:

```python
file = univpm1.drive.CreateFile({'id': "13nYjSBX7FmSUFnBcTuT6hKsFswXhqcpzpioei5r2eRQ"})
file.FetchMetadata()
print(json.dumps(file, indent=4))
```

```json
{
    "id": "13nYjSBX7FmSUFnBcTuT6hKsFswXhqcpzpioei5r2eRQ",
    "kind": "drive#file",
    "userPermission": {
        "id": "me",
        "type": "user",
        "role": "owner",
        "kind": "drive#permission",
        "selfLink": "https://www.googleapis.com/drive/v2/files/13nYjSBX7FmSUFnBcTuT6hKsFswXhqcpzpioei5r2eRQ/permissions/me",
        "etag": "\"_QYtqCZlWmDSoT348qPSAnswYHU\"",
        "pendingOwner": false
    },
    "selfLink": "https://www.googleapis.com/drive/v2/files/13nYjSBX7FmSUFnBcTuT6hKsFswXhqcpzpioei5r2eRQ",
    "ownerNames": [
        "user1"
    ],
    "lastModifyingUserName": "user1",
    "editable": true,
    "writersCanShare": true,
    "mimeType": "application/vnd.google-apps.document",
    "exportLinks": {
        "application/rtf": "https://docs.google.com/feeds/download/documents/export/Export?id=13nYjSBX7FmSUFnBcTuT6hKsFswXhqcpzpioei5r2eRQ&exportFormat=rtf",
        "application/vnd.oasis.opendocument.text": "https://docs.google.com/feeds/download/documents/export/Export?id=13nYjSBX7FmSUFnBcTuT6hKsFswXhqcpzpioei5r2eRQ&exportFormat=odt",
        "text/html": "https://docs.google.com/feeds/download/documents/export/Export?id=13nYjSBX7FmSUFnBcTuT6hKsFswXhqcpzpioei5r2eRQ&exportFormat=html",
        "application/pdf": "https://docs.google.com/feeds/download/documents/export/Export?id=13nYjSBX7FmSUFnBcTuT6hKsFswXhqcpzpioei5r2eRQ&exportFormat=pdf",
        "application/epub+zip": "https://docs.google.com/feeds/download/documents/export/Export?id=13nYjSBX7FmSUFnBcTuT6hKsFswXhqcpzpioei5r2eRQ&exportFormat=epub",
        "application/zip": "https://docs.google.com/feeds/download/documents/export/Export?id=13nYjSBX7FmSUFnBcTuT6hKsFswXhqcpzpioei5r2eRQ&exportFormat=zip",
        "application/vnd.openxmlformats-officedocument.wordprocessingml.document": "https://docs.google.com/feeds/download/documents/export/Export?id=13nYjSBX7FmSUFnBcTuT6hKsFswXhqcpzpioei5r2eRQ&exportFormat=docx",
        "text/plain": "https://docs.google.com/feeds/download/documents/export/Export?id=13nYjSBX7FmSUFnBcTuT6hKsFswXhqcpzpioei5r2eRQ&exportFormat=txt"
    },
    "parents": [
        {
            "selfLink": "https://www.googleapis.com/drive/v2/files/13nYjSBX7FmSUFnBcTuT6hKsFswXhqcpzpioei5r2eRQ/parents/1IETDYYj23PgGpInZofa9MyANyBlOoiyh",
            "id": "1IETDYYj23PgGpInZofa9MyANyBlOoiyh",
            "isRoot": false,
            "kind": "drive#parentReference",
            "parentLink": "https://www.googleapis.com/drive/v2/files/1IETDYYj23PgGpInZofa9MyANyBlOoiyh"
        }
    ],
    "thumbnailLink": "https://docs.google.com/feeds/vt?gd=true&id=13nYjSBX7FmSUFnBcTuT6hKsFswXhqcpzpioei5r2eRQ&v=2&s=AMedNnoAAAAAYuhkOb34StnrPTCnQ3wcBSVvE9alyRKz&sz=s220",
    "appDataContents": false,
    "iconLink": "https://drive-thirdparty.googleusercontent.com/16/type/application/vnd.google-apps.document",
    "shared": false,
    "lastModifyingUser": {
        "displayName": "user1",
        "kind": "drive#user",
        "isAuthenticatedUser": true,
        "permissionId": "16663488747090924602",
        "emailAddress": "user1@gmail.com",
        "picture": {
            "url": "https://lh3.googleusercontent.com/a/default-user=s64"
        }
    },
    "owners": [
        {
            "displayName": "user1",
            "kind": "drive#user",
            "isAuthenticatedUser": true,
            "permissionId": "16663488747090924602",
            "emailAddress": "user1@gmail.com",
            "picture": {
                "url": "https://lh3.googleusercontent.com/a/default-user=s64"
            }
        }
    ],
    "copyable": true,
    "etag": "\"MTY1ODg2MzYzNDQ2Nw\"",
    "alternateLink": "https://docs.google.com/document/d/13nYjSBX7FmSUFnBcTuT6hKsFswXhqcpzpioei5r2eRQ/edit?usp=drivesdk",
    "embedLink": "https://docs.google.com/document/d/13nYjSBX7FmSUFnBcTuT6hKsFswXhqcpzpioei5r2eRQ/preview?ouid=107468266404355780381",
    "fileSize": "1024",
    "copyRequiresWriterPermission": false,
    "spaces": [
        "drive"
    ],
    "title": "file1",
    "labels": {
        "viewed": true,
        "restricted": false,
        "starred": false,
        "hidden": false,
        "trashed": false
    },
    "explicitlyTrashed": false,
    "createdDate": "2022-07-26T19:13:25.265Z",
    "modifiedDate": "2022-07-26T19:27:14.467Z",
    "modifiedByMeDate": "2022-07-26T19:27:14.467Z",
    "lastViewedByMeDate": "2022-07-26T19:27:14.467Z",
    "markedViewedByMeDate": "1970-01-01T00:00:00.000Z",
    "quotaBytesUsed": "1024",
    "version": "13",
    "capabilities": {
        "canEdit": true,
        "canCopy": true
    }
}
```

We can see that `parents` is an array. So by design it should be possible for a file to have multiple parents.

But as we saw in https://github.com/iterative/PyDrive2/pull/188#discussion_r921783275, at least in my own drive, is not possible to have a file with multiple parents.

But if in other types of drives is possible for a file to have multiple parents, the line `file1["parents"] = [{"id": file2_parent_id}]` will be a big bug.

shcheklein commented:
@simone-viozzi multiple parents are not possible since 2020 I think https://developers.google.com/drive/api/guides/ref-single-parent#:~:text=Beginning%20Sept.,exactly%20one%20parent%20folder%20location.

shcheklein commented on pydrive2/apiattr.py:
```diff
@@ -94,12 +94,24 @@ def GetChanges(self):
         dirty = {}
         for key in self:
             if self.metadata.get(key) is None:
+                print(
```
we can remove this now, I think

shcheklein commented on pydrive2/fs/spec.py:
```diff
@@ -542,6 +542,35 @@ def cp_file(self, lpath, rpath, **kwargs):
+
+        file1 = self.client.CreateFile({"id": file1_id})
+
+        file1.FetchMetadata(fields="title,parents")
```
https://github.com/iterative/PyDrive2/pull/223 should fix this, please rebase and let's try to simplify this

shcheklein commented on pydrive2/fs/spec.py:
```diff
@@ -542,6 +542,35 @@ def cp_file(self, lpath, rpath, **kwargs):
+        file2_name = posixpath.basename(path2)
+
+        file1_parent = self._parent(path1)
+        file2_parent = self._parent(path2)
```
Looked at the code - so it can be empty I think (if `path2` is `root` for example). Need to check that everything works in that case.

simone-viozzi replied:
it does work if you specify the name of the target:

```python
fs.mv('root/tmp/a.pdf', 'root/b.pdf')
```

but you do:

```python
fs.mv('root/tmp/a.pdf', 'root/)
```

it will throw an error.

shcheklein replied:
it should be working though, right? It should be close to `mv` UNIX semantics.

simone-viozzi replied:
Fixed.
Now it works with or without the file name, and with or without the leading slash:

```python
fs.mv('root/tmp/b.pdf', 'root/')
fs.mv('root/tmp/b.pdf', 'root')
fs.mv('root/tmp/b.pdf', 'root/a.pdf')
```

simone-viozzi commented:
The way it works right now you always need to specify the target name. 
@shcheklein is this the same behavior as before?

 Also the recursive part does not need almost any implementation, because on gdrive if you move a folder the content is moved automatically.

The max-depth part will be difficult though. Because we need to implement the recursive logic.

simone-viozzi commented:
@shcheklein how you want to handle `recursive` and `maxdepth`?
Right now is recursive by default, I think we can add a check so if the source is a folder and `recursive=False` it raises an exception. What do you think?

Regarding `maxdepth` I think is ok to leave it not implemented.

simone-viozzi commented:
Hey @shcheklein could you please review this PR? 
I think it's ready to be merged

shcheklein commented on pydrive2/fs/spec.py:
```diff
@@ -542,6 +542,43 @@ def cp_file(self, lpath, rpath, **kwargs):
+    @_gdrive_retry
+    def mv(self, path1, path2, recursive=False, maxdepth=None, **kwargs):
+
+        if maxdepth:
```
Let's do `is not None`. It should be a bit more precise

shcheklein commented on pydrive2/fs/spec.py:
```diff
@@ -542,6 +542,43 @@ def cp_file(self, lpath, rpath, **kwargs):
+        if maxdepth:
+            raise NotImplementedError("Max depth move is not supported")
+
+        if self.isdir(path1) and not recursive:
```
Q: how do we handle recursive after all? Why do we need this extra check. Is it only as a safeguard to check the intention? I'm asking since even the `isdir` check can be expensive (extra call).

simone-viozzi replied:
If you move a folder, the content is moved with it. So is kinda always recursive.

The check is there to avoid that a user wanted to move a file (`recursive=False`) but by accident the source is a directory.

We could also remove `recursive` and always assume that the user want to move the entire folder. This way, there is no need to check if the source is a directory.

shcheklein replied:
I would say yes (to avoid that extra API call). Also data does not disappear anywhere, it's not that dangerous.

how does the local filesystem fsspec implementation approach this? 

unix `mv` doesn't require the recursive flag afair

simone-viozzi replied:
I did that to not modify the default method signature of `ffspec`:
https://github.com/fsspec/filesystem_spec/blob/2633445fc54797c79a7ac96d213bd24dfbdfcdc2/fsspec/spec.py#L917-L920
In this case, the flags are passed down to copy.


But yes, if it's allowed to change the signature, we can also remove `recursive` and the `isDir` check.

shcheklein replied:
I would just ignore it. We are not allowed to change the default.

We can also check still how they implement the local fs. Is it unix way or it check the recursive flag.

simone-viozzi replied:
> We can also check still how they implement the local fs.  Is it unix way or it check the recursive flag.

Where?

shcheklein commented on pydrive2/fs/spec.py:
```diff
@@ -542,6 +542,43 @@ def cp_file(self, lpath, rpath, **kwargs):
+                "Cannot move a directory without recursive flag"
+            )
+
+        file1_name = posixpath.basename(path1)
```
Let's name things `src` / `dst`. Since they are not files in all cases and it's more descriptive.

shcheklein commented on pydrive2/fs/spec.py:
```diff
@@ -542,6 +542,43 @@ def cp_file(self, lpath, rpath, **kwargs):
+
+        file1.Upload()
+
+        # TODO need to invalidate the cache for the old path
```
Let's move it up, before the call, create a ticket in GH and put a link to it pls

simone-viozzi replied:
I created the issue #232 and referenced it in the comment.

shcheklein replied:
thanks!

shcheklein reviewed (changes requested):
Hey, some stylistic changes. Good stuff overall. I'm not sure how recursive is handled, also we need to add test, would be great to count the number of api calls in those tests also

simone-viozzi commented:
@shcheklein there is already a test:
https://github.com/iterative/PyDrive2/blob/688459b9c70c2e398b6dcf6178217f80ef76ee78/pydrive2/test/test_fs.py#L90

shcheklein commented:
> @shcheklein there is already a test:
> 
> https://github.com/iterative/PyDrive2/blob/688459b9c70c2e398b6dcf6178217f80ef76ee78/pydrive2/test/test_fs.py#L90

Q: How did it work before if mv was not implemented?

simone-viozzi commented:
@shcheklein regarding the recursive behavior, there is no actual need of recursion. 
If we change the parent of a folder, all the subfolders will still point to the same id. So they will move along with their parent.

So it is possible to move an entire folder structure with just one call.

simone-viozzi commented:
> > @shcheklein there is already a test:
> > https://github.com/iterative/PyDrive2/blob/688459b9c70c2e398b6dcf6178217f80ef76ee78/pydrive2/test/test_fs.py#L90
> 
> Q: How did it work before if mv was not implemented?

Before `mv` was implemented with `copy` + `rm`:
https://github.com/fsspec/filesystem_spec/blob/2633445fc54797c79a7ac96d213bd24dfbdfcdc2/fsspec/spec.py#L917-L920

shcheklein commented:
> @shcheklein regarding the recursive behavior, there is no actual need of recursion.

okay, let's remove / ignore it for now

simone-viozzi commented:
I removed the `recursive` flag. If a user use it, it will go into the `kwargs` and it will be ignored.

shcheklein commented:
Thanks @simone-viozzi !
